personalities/sctfe/storage/bbolt/dedup.go

```diff
@@ -0,0 +1,179 @@
+		} else if dedupB != nil && sizeB == nil {
+			return fmt.Errorf("inconsistent deduplication storage state, %q is not nil but %q is nil", dedupBucket, sizeBucket)
+		} else {
+			klog.V(2).Infof("NewStorage: found pre-existing %q and %q buckets.", dedupBucket, sizeBucket)
+		}
+		return nil
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("error initializing buckets: %v", err)
+	}
+
+	return s, nil
+}
+
+// Add inserts entries in the dedup bucket and updates the size bucket if need be.
+//
+// If an entry is already stored under a given key, Add does not update it, even if the new value
+// is different.
+// The context is here for consistency with interfaces, but isn't used by BBolt.
+func (s *Storage) Add(_ context.Context, kvs []dedup.KV) error {
+	for _, kv := range kvs {
+		err := s.db.Update(func(tx *bolt.Tx) error {
+			db := tx.Bucket([]byte(dedupBucket))
+			sb := tx.Bucket([]byte(sizeBucket))
+			sizeB := sb.Get([]byte("size"))
+			if sizeB == nil {
+				return fmt.Errorf("can't find log size in bucket %q", sizeBucket)
+			}
+			size := btoi(sizeB)
+
+			if old := db.Get(kv.K); old != nil {
+				klog.V(3).Infof("Add(): bucket %q already contains an entry for %q, not updating", dedupBucket, hex.EncodeToString(kv.K))
+			}
+			if err := db.Put(kv.K, itob(kv.V)); err != nil {
+				return err
+			}
+			// sizeB is indexes from 1 since it's a size, kv.K from 0.
+			// Therefore, if they're equal, kv is a new entry.
+			if size == kv.V {
+				klog.V(3).Infof("Add(): updating deduped size to %d", size+1)
+				if err := sb.Put([]byte("size"), itob(size+1)); err != nil {
+					return err
+				}
+			}
+			return nil
+		})
+		if err != nil {
+			return fmt.Errorf("b.Put(): error writting leaf index %d: err", kv.V)
+		}
+	}
+	return nil
+}
+
+// Get reads entries from the dedup bucket.
+//
+// The context is here for consistency with interfaces, but isn't used by BBolt.
+func (s *Storage) Get(_ context.Context, leafID []byte) (uint64, bool, error) {
+	var idx []byte
+	_ = s.db.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte(dedupBucket))
+		v := b.Get(leafID)
+		if v != nil {
+			idx = make([]byte, 8)
+			copy(idx, v)
+		}
+		return nil
+	})
+	if idx == nil {
+		return 0, false, nil
+	}
+	return btoi(idx), true, nil
+}
+
+// LogSize reads the latest entry from the size bucket.
+func (s *Storage) LogSize() (uint64, error) {
+	var size []byte
+	err := s.db.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte(sizeBucket))
+		v := b.Get([]byte("size"))
+		if v != nil {
+			size = make([]byte, 8)
+			copy(size, v)
+		}
+		return nil
+	})
+	if err != nil {
+		return 0, fmt.Errorf("error reading from %q: %v", sizeBucket, err)
+	}
+	if size == nil {
+		return 0, fmt.Errorf("can't find log size in bucket %q", sizeBucket)
+	}
+	return btoi(size), nil
+}
+
+// itob returns an 8-byte big endian representation of idx.
+func itob(idx uint64) []byte {
+	b := make([]byte, 8)
+	binary.BigEndian.PutUint64(b, uint64(idx))
+	return b
+}
```

Comment:
Might be able to remove this by using `binary.BigEndian.AppendUint64(nil, N)` directly inline?

Can inline `btoi` too since it's only one line?

Reply:
Done. I find that they make the code easier to read, so I'll leave them around.